src/app/upgrade.cpp

```diff
@@ -53,17 +52,10 @@ namespace
             if (!newData.isEmpty() || oldData.isEmpty())
                 return;
 
-            QFile file(savePath);
-            if (!file.open(QIODevice::WriteOnly))
+            const std::optional<QString> result {Utils::Fs::saveToFile(savePath, oldData)};
```

Comment:
Seems you didn't understand my opinion: Error message isn't direct product of function so I dislike such a signature.

Reply:
>Seems you didn't understand my opinion: Error message isn't direct product of function so I dislike such a signature.

Damn!
Now I see that my previous comment about it is lost.

Reply:
>Error message isn't direct product of function so I dislike such a signature.

IMO you got the wrong viewpoint of `std::optional<T>` usage.
First your first part is right: "Error message isn't direct product of function". The second part "such a signature", `std::optional<T>` encapsulate a type `T` that is returned on error. Here `T` is just `QString` because there are no other interesting fields to the caller, if there are we can make another `struct` and returns it.
Or if you insist we can define a "error" type just to encapsulate the `QString`:
```c++
struct SaveFileError
{
    QString errorMessage;
}

std::optional<SaveFileError> Utils::Fs::saveToFile(path, data)
{
    // ...
}
```

Reply:
>First your first part is right: "Error message isn't direct product of function".

My comment can be generalized as follows:
"Error isn't direct product of function". So `std::optional<SaveFileError> Utils::Fs::saveToFile(path, data) ` doesn't look better to me.

Reply:
>"Error isn't direct product of function". 

Then I suppose returning `bool` isn't acceptable too.

Reply:
Now I provided a wrapper class for it. This would be the last attempt on this issue.

Reply:
Sorry for the late response. I'm very busy in real life right now. Even when I'm writing this comment I am in a hurry, so there may be something missing in my comment.

>Then I suppose returning bool is unacceptable too. Only throwing exceptions is allowed, right? If that is true then it would be too narrow/shallow IMO.

In an idealized model, I think yes. There we would always have an explicit division into the result of the function, the side effect of the function, and exception handling during the execution of the function. But it is impossible to strictly follow such a model in the real world for various reasons (in some cases it is inconvenient, in some cases it would add unnecessary complication, etc.). Therefore, we have to use other ways to provide the operation status. I just want us to use one of the well-known / widely used methods, if one of these completely suits us and that we have some consistency within our code, i.e. that we do not use different methods in similar cases (of course, if we are not talking about third-party things).
I myself am not in favor of using exceptions everywhere, but in some situations they clearly win. In the case under discussion, exceptions do not seem to be strictly required, so you can go the other way. But I don't like the ways you're trying to go here at all. They look illogical to me. You violate some common comprehension, without providing any tangible advantages. I have already mentioned the previous one. In the new one, you created a class that performs its only work in its constructor. This also looks pointless/illogical to me.
Why not just use well known way like the following?
```c++
bool saveToFile(const QString &path, const QByteArray &data, QString *error);
```

Then the operation status can be checked in a familiar and understandable way:
```c++
QString error;
if (!saveToFile(path, data, &error))
{
    printError(error);
}
```
And you can easily even ignore the details of the error, if you are not interested in it:
```c++
if (saveToFile(path, data))
{
    // do something
}
```

Of course, you can go some newfangled way and wrap the entire result of the function execution in some object, but then in order to look logical, it should represent exactly the "aggregated result" of the function execution, and not an (optional) error, as you did last time. This "aggregated result" should be castable to the boolean type so that we can use it in the usual way (like above):
```c++
const Result res = saveToFile(path, data);
if (res.errored)
{
    printError(res.errorMessage);
}
// or
if (saveToFile(path, data))
{
    // do something
}
```

Reply:
>There we would always have an explicit division into the result of the function, the side effect of the function, and exception handling during the execution of the function.

Really we have two classes of subroutines:
1. Function (in its essence), i.e. a subroutine that results in a return value calculated based on the received parameters (e.g. any kind of arithmetic, logic, conversion subroutines);
2. a subroutine that results in its side effect.

I believe that it is clear without any additional discussion that the first class of subroutines should return exactly the result of the function in its essence (otherwise many familiar and convenient things will be broken). Error handling in them should be done either with the help of exceptions, or in some other way that does not interfere with their regular use (it should be possible to use them as part of complex expressions as parameters of other functions or operators). So here we are discussing the second class of subroutines.

Reply:
>Sorry for the late response. I'm very busy in real life right now. Even when I'm writing this comment I am in a hurry, so there may be something missing in my comment.

OK, we can slow it down.

>I have already mentioned the previous one. In the new one, you created a class that performs its only work in its constructor. This also looks pointless/illogical to me.

As long as the class interface/usage looks legit, the rest is implementation details and can be overlooked IMO.

>bool saveToFile(const QString &path, const QByteArray &data, QString *error);

I also do not favor function output parameters, if the function has to show some results it should return them all via the usual way (the `return` statement). IMO output parameters are just abusing the reference type to achieve side-effects. It also makes a function less "pure" (as in pure functions, see below).

>Of course, you can go some newfangled way and wrap the entire result of the function execution in some object, but then in order to look logical, it should represent exactly the "aggregated result" of the function execution, and not an (optional) error

I think you are contradicting your self conceptually, whatever the error handing mechanism is, be it your suggested `Result`, `std::optional<T>`, exception `RuntimeError`, they are all doing the same thing: reporting back what happened in the function. They are all the same in "easily ignored if intended" and only differs in actual error handling code (`if()`, `catch`).
I do not understand why would you think using `std::optional` is inappropriate. Here is the first paragraph of https://en.cppreference.com/w/cpp/utility/optional:
> The class template std::optional manages an optional contained value, i.e. a value that may or may not be present.

Which fit exactly into what our function would return: if `saveToFile()` fails then an additional error message is produced and a `bool` will signal it. If `saveToFile()` succeeds then all good, no error message and the `bool` will also signal it.

I'll also add that people (not just me) really want to make use of error handing via typing system instead of exceptions: see [boost.outcome](https://www.boost.org/doc/libs/develop/libs/outcome/doc/html/index.html) and the proposed [std::expect](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p0323r10.html).

>This "aggregated result" should be castable to the boolean type so that we can use it in the usual way (like above):

It sounds most likely reinventing `std::optional<T>`...

>Really we have two classes of subroutines:

This is getting a little theoretical... The first class is called [pure functions](https://en.wikipedia.org/wiki/Pure_function), basically we want every function to be this kind (or as much as possible). However error handling mechanisms are still very much the same (for all classes actually). Consider a pure function that will perform "divide by zero", you can let `/` operator throw default exceptions or wrap the result with a `Result` type (as people want safe division in their context).

>So here we are discussing the second class of subroutines. 

Yes. I'll add that *most* of the code in our codebase (or even in C++) are in the second class, which error handling methods are still very much the same as everywhere.

---
tldr. currently seems the only acceptable for both party is `const Result res = saveToFile(path, data);`, then so be it, I'll update it later if there are no new opinions.

Reply:
>currently seems the only acceptable for both party is `const Result res = saveToFile(path, data);`, then so be it, I'll update it later if there are no new opinions.

Please wait until we finished our discussion. As I said above I need more time to comment on this topic currently. 
If you think that this Issue should be solved as soon as possible, then you can return the original version of PR, which does not have any `saveToFile()` helper. Then we can merge it and continue the discussion elsewhere.

Reply:
>I do not understand why would you think using `std::optional` is inappropriate.

IMO, it can be appropriate if T is Result of function (so empty `optional` means "cannot produce result for some reason") and you don't need to provide additional info in exceptional case (or you provide it separately). IMO, this is only appropriate in cases where the inability to provide a specific result is not an error, for example, when we have some set of settings successfully read from a file, and if some specific option is missing, then this is not an error and we just use the default value.

>Which fit exactly into what our function would return: if `saveToFile()` fails then an additional error message is produced and a bool will signal it. If `saveToFile()` succeeds then all good, no error message and the bool will also signal it. 

This is exactly what looks illogical/unusual/inconsistent.  Expression `if (saveToFile(path, data))` is perceived as "if save to file was successful", but your suggestion (`optional<Error>`) behaves the other way around, which makes the code confusing, especially since in other similar cases we get `true` precisely as a result of the successful execution of the operation.
>>This "aggregated result" should be castable to the boolean type so that we can use it in the usual way (like above):
>
>It sounds most likely reinventing `std::optional<T>`

I would say that this is similar to `Boost.Outcome`, since the return value itself is not optional, it always contains some information about the operation status, and the main result, if it is available.

Reply:
>This is getting a little theoretical... The first class is called pure functions, basically we want every function to be this kind (or as much as possible).

How can we want this? How can we make `pure` function that isn't inherently `pure`? How can you imagine `saveToFile()` as a pure function, if its meaning is to produce a side effect (writing data to disk)?
>Consider a pure function that will perform "divide by zero", you can let `/` operator throw default exceptions or wrap the result with a `Result` type (as people want safe division in their context).

Wrapping result of `pure` function with a compound `Result` type can make the function `impure` in general case.

But it seems that this aspect of the discussion is least relevant.

I would still like to insist on the following:
1. We should not prohibit "exception handling" (`try-catch`) in cases where it has an advantage;
2.We should strive to unify another way of error handling, i.e. we should choose one method and use it, and not use something different each time.

Reply:
>This is exactly what looks illogical/unusual/inconsistent. Expression if (saveToFile(path, data)) is perceived as "if save to file was successful", but your suggestion (optional<Error>) behaves the other way around, which makes the code confusing, especially since in other similar cases we get true precisely as a result of the successful execution of the operation.

In our case here, I agree that `if (saveToFile(path, data))` expression should not be used because it leads to said confusion. However I would still argue it is not due to inappropriate usage of std::optional, instead, it is due to misunderstanding of the function return (and the underlying meaning of the type signature) because as said before it means that `the "error message" field presence is optional, not the concept of error is optional`.
You are insisting on only returning valid results while here it is the exact opposite: it is returning an error message and I would say `std::optional` itself doesn't regulate what should be or not, "optional" doesn't imply meaning of "valid" or "results", it is just optional. IMO usage largely depends on the context and insisting on only one form isn't reasonable.
The use of having error message within `std::optional` also does not violates ["When to use Optional" from boost:](https://www.boost.org/doc/libs/1_77_0/libs/optional/doc/html/boost_optional/tutorial/when_to_use_optional.html)
>It is recommended to use optional<T> in situations where there is exactly one, clear (to all parties) reason for having no value of type T, and where the lack of value is as natural as having any regular value of T.

My interpretation: no error = no error code/message thus no `T` exist.

>How can we want this? How can we make pure function that isn't inherently pure? How can you imagine saveToFile() as a pure function, if its meaning is to produce a side effect (writing data to disk)?

It won't be feasible in imperative programming languages but functional programming languages have their way to deal with it, I'll let you research it yourself.

>Wrapping result of pure function with a compound Result type can make the function impure in general case.

I don't think your statement is universally true as returning a `Result` type (if correctly coded) won't violate any properties in the mentioned wiki link.

>1. We should not prohibit "exception handling" (try-catch) in cases where it has an advantage;

I agree, it is proven to be useful at https://github.com/qbittorrent/qBittorrent/blob/master/src/base/bittorrent/torrentcreatorthread.cpp#L104

>2. We should strive to unify another way of error handling, i.e. we should choose one method and use it, and not use something different each time.

I don't agree much on this or at least with the current wording. IMO Each way error handling has its own merits and we shouldn't limit ourselves on only one. Personally I would lean more on no-exception handling for general cases (as it won't disrupt code flow in contrast to exceptions) and only use exceptions when it provide benefits over the others, such as simplifying the handling procedures  & automatic clean up of objects or 3rd party code requirements...

Reply:
>It won't be feasible in imperative programming languages but functional programming languages have their way to deal with it, I'll let you research it yourself.

We're talking about C++, aren't we? So even if this aspect of the discussion is more theoretical, I would prefer to limit its scope to those that can somehow affect the practical aspects of qBittorrent development (functional programming languages, etc. are not included there).

>You are insisting on only returning valid results while here it is the exact opposite

Anyway having "error message" (even optional) as function return type looks ugly to me. It looks as if the purpose of the function is to create error messages.
>The use of having error message within std::optional also does not violates "When to use Optional" from boost:
>
>>It is recommended to use optional in situations where there is exactly one, clear (to all parties) reason for having no value of type T, and where the lack of value is as natural as having any regular value of T.

IMO, `optional<ErrorMessage>` violates "the lack of value is as natural as having any regular value of T" as well as `optional<Result type>` in case `nullopt` means function failure (as opposite to my example above).
>IMO Each error handling method has its own merits and we shouldn't limit ourselves on only one.

I mean we shouldn't use different methods in similar cases.
>Personally I would lean more on no-exception handling for general cases (as it won't disrupt code flow in contrast to exceptions) and only use exceptions when it provide clear benefits over the others

This is what I said in par. 1 above:
>We should not prohibit "exception handling" (try-catch) in cases where it has an advantage;

Reply:
>`if (saveToFile(path, data))` expression should not be used because it leads to said confusion. 

If we consider the interface of `saveToFile()` you proposed, then yes. But I repeat once again that I do not consider it suitable/correct.
You seem to be thinking narrowly about this. You are not using this particular function in this context. But it is natural for this kind of functions (which are expected to return the operation status) to be used in such a context (and, by the way, `Boost.Outcome` and `std::expected` are designed this way):
```c++
if (saveToFile(path, data))
{
    // do something in case of successful saving
}

// or

if (!saveToFile(path, data))
{
    // do something in case of failed saving
}

// or if you need result

const auto result = saveToFile(path, data);
if (!result)
{
    message(result.error());
}

// or in c++17

if (const auto result = saveToFile(path, data); !result)
{
    message(result.error());
}
```

The fact that this particular function differs in that the type of its main result is `void` does not mean that its interface should be semantically different from other similar functions that have a `non-void` type of main result. `std::expected` has good [clarification](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p0323r10.html#void) of this case:
>The type `expected<void, E>` means "nothing is expected, but an error could occur".

Reply:
>So even if this aspect of the discussion is more theoretical, I would prefer to limit its scope to those that can somehow affect the practical aspects of qBittorrent development (functional programming languages, etc. are not included there).

I already said: you can do your own research, google is for everyone.

>I mean we shouldn't use different methods in similar cases.

Are you sure you will able to unify all (similar) cases to use only one error handling method? and are you sure that won't cause greater inconvenience? maybe in an idealized world...

> > Personally I would lean more on no-exception handling for general cases (as it won't disrupt code flow in contrast to exceptions) and only use exceptions when it provide clear benefits over the others
> 
> This is what I said in par. 1 above:
> 
> > We should not prohibit "exception handling" (try-catch) in cases where it has an advantage;

Yes but in the context at hand, I don't really see an advantage for using exceptions.

>and, by the way, Boost.Outcome and std::expected are designed this way

So what is the conclusion? Do you accept usage of Boost.Outcome and/or std::expected for the function?

Reply:
>Are you sure you will able to unify all (similar) cases to only one error handling method? and are you sure that won't cause greater inconvenience? maybe in an idealized world...

You seem to be complicating this... You don't have to unify the cases so that they become similar, because they are either similar or not initially. So we just choose which method to use in a particular kind of cases. Moreover, (preliminary) decisions are already emerging from this discussion.
You insist on avoid using exceptions if they do not bring clear advantages in a particular case. Well, so be it.
Then it remains to provide an "error descriptor" either through the output parameter of the function, or through the return value (well, formally there are still other ways, such as a global variable, but I think that you also consider this unacceptable). You are against using output parameters for this purpose. I agree.
Only the return value remains. Fine, let's use it. I only insist that this be done uniformly for any type of the "main result" of the function, including the `void` (so there is always clear distinction between main result and error descriptor).

Sidenote:
I will also point out again that there may be functions (such as I mentioned above) for which the inability to return a result under some conditions is not an erroneous case (so they do not fall under the error handling discussed here).

>So what is the conclusion? Do you accept usage of `Boost.Outcome` and/or `std::expected` for the function?

`std::expected` is unreleased, IIRC.
So we can use `Boost.Outcome` unless you want to provide custom implementation.

Reply:
However, `Boost.Outcome` uses `std::error_code` as the default error type. I think it's not a good idea to mess with it, so we can use custom error types (maybe just QString if all we need is error message).

Reply:
>Only the return value remains. Fine, let's use it. I only insist that this be done uniformly for any type of the "main result" of the function, including the void (so there is always clear distinction between main result and error descriptor).

OK.

>std::expected is unreleased, IIRC.
So we can use Boost.Outcome unless you want to provide custom implementation.

There are some single header implementations on the net that we can consider:
https://github.com/martinmoene/expected-lite
https://github.com/TartanLlama/expected
One concern is that `std::expected` hasn't finalized yet so the implementations might diverge from the final version. However I'm inclined to think that the differences won't be too significant to handle.
Another concern is those implementations quality. There are some issues (on their issue tracker) which I'm not sure whether it will have impact on us.

A related question: which do we prefer between `Boost.Outcome` and finalized `std::expected`?
Personally I lean to the latter.

>However, Boost.Outcome uses std::error_code as the default error type. I think it's not a good idea to mess with it, so we can use custom error types (maybe just QString if all we need is error message).

Agree.

Reply:
>A related question: which do we prefer between `Boost.Outcome` and finalized `std::expected` (when that day comes)?
Personally I lean to the latter.

I would also prefer `std::expected`, all other things being equal. So we can use `expected-lite` until then.
>Another concern is those implementations quality. There are some issues (on their issue tracker) which I'm not sure whether it will have impact on us.

I think we could even use `Boost.Outcome` until `std::expected` is released. I don't think the transition will be more complicated than applying "Find and Replace".

Reply:
PR updated, I went with `expected-lite` and it works fine on my system.